cibuildwheel/linux.py

```diff
@@ -129,72 +117,34 @@ def build(project_dir, package_name, output_dir, test_command, test_requires, be
-            command,
-            stdin=subprocess.PIPE,
-            universal_newlines=True,
-        )
-
-        # It will take a bit for docker to get the container up and
-        # running.  Keep trying to copy in the project directory until
-        # it succeeds.  There is a timeout to avoid spinning forever.
-
-        timeout = 180
-        start = monotonic.monotonic()
-        while True:
-            time.sleep(1)
-            command = [
-                'docker',
-                'cp',
-                '{}/.'.format(os.path.abspath(project_dir)),
-                '{}:/project'.format(container_name),
-            ]
-            print('docker command: {}'.format(command))
-            if subprocess.call(command) == 0:
-                break
-
-            if monotonic.monotonic() - start > timeout:
-                raise DockerRunTimeoutError(
-                    'Unable to successfully copy project directory'
-                    ' within {} seconds'.format(timeout)
-                )
+        def run_docker(command, stdin_str=None):
+            print('docker command: docker {}'.format(' '.join(map(shlex_quote, command))))
+            if stdin_str is None:
+                subprocess.check_call(['docker'] + command)
+            else:
+                process = subprocess.Popen(['docker'] + command,
+                                           stdin=subprocess.PIPE, universal_newlines=True)
+                try:
+                    process.communicate(stdin_str)
+                except KeyboardInterrupt:
+                    process.kill()
+                    process.wait()
+                if process.returncode != 0:
+                    raise subprocess.CalledProcessError(process.returncode, process.args)
 
+        container_name = 'cibuildwheel-{}'.format(uuid.uuid4())
         try:
-            docker_process.communicate(bash_script)
-        except KeyboardInterrupt:
-            docker_process.kill()
-            docker_process.wait()
-
-        command = [
-            'docker',
-            'cp',
-            '{}:/output/.'.format(container_name),
-            os.path.abspath(output_dir),
-        ]
-        print('docker command: {}'.format(command))
-        subprocess.check_call(command)
-
-        command = [
-            'docker',
-            'rm',
-            '-v', container_name,
-        ]
-        print('docker command: {}'.format(command))
-        subprocess.check_call(command)
-
-        if docker_process.returncode != 0:
+            run_docker(['create',
+                        '--env', 'CIBUILDWHEEL',
+                        '--name', container_name,
+                        '-i',
+                        '-v', '/:/host', # ignored on Circle
+                        docker_image, '/bin/bash'])
+            run_docker(['cp', os.path.abspath(project_dir) + '/.', container_name + ':/project'])
+            run_docker(['start', '-i', '-a', container_name], stdin_str=bash_script)
```

Comment:
How about we skip the communicate stuff?  Have a real .sh file instead of a string (which could happen regardless) and copy it into, say, `/cibuildwheel.sh` and then just have docker run it.  Then it can always `check_call()`.

Reply:
I read your comment on this in the other conversation, and I actually tried out some stuff, but it indeed seems that the only way would be to create a file, copy it into the image, and run it from there. And that seems +- as involved as the current solution? Plus, creating that file and copying it seems ... I don't know. I think a `stdin` solution is quite nice (though admittedly, the `communicate` step isn't really).

Reply:
@joerick Any thoughts?

Reply:
I think that regardless the bash script should probably be stored in the repository as a .sh file.  If the present `.communicate()` approach were kept then you'd `with open('the_bash.sh') as f: script = f.read()` then `.communicate()` the `script` string as we do now.

Sure, there are other steps, but...

Reply:
Hmmm, but what about the `CIBW_BEFORE_BUILD` etc, that is inserted in the script template? Are you suggesting to include those in other ways? I guess environment variables could work as well, to insert custom steps. Before, I thought you meant writing the script from Python string to a temporary file, each time, so I hadn't considered this possibility.

Personally, I do actually quite like having all code related to building on Linux in one file, just like it is for macOS and Windows (even though, I admit, it looks kind of ugly to have a full bash script inside a Python string literal).

Reply:
Alrighty, you got me...  I skipped over that detail since I was talking abstractly instead of trying to implement my idea.  Still, it could be the same thing it is now but in a file.  Then you get syntax highlighting for bash etc (other than issues caused by the substitution markers).  But sure, I'll back off a bit.  I might still rather write to a temp file but...  whatever, I'll back off.  :]

In another direction, don't we get Python in the container?  Couldn't this be a Python script and it could load a config file that we write?

Obviously none of this is needed, just sharing thoughts at this point.

Reply:
In `communicate` vs temp file, I'm very slightly in favour of communicate, just seems neat not to involve an extra file somehow!

Yeah, the script is bash since I thought it was really confusing to write a python script inside a string literal inside a python file!! So just wrote it in bash to get it done. It would be cool if somehow it was a fully fledged python script though. Maybe then there would be potential to unify the Linux/Mac/windows divide? :0

Reply:
@YannickJadoul there was no negative connotation intended in my message.  :]  You honestly caught me pushing an idea that I hadn't thought through and you had found an issue with.  Anyways, I _think_ that _maybe_ a good way would be a Python script with a CLI (probably small enough to comfortably do with stdlib and 2/3 compatibility) and we pass in the info via the command.  No config file to write and copy, no template to fill out.  Just copy the exact Python script over to the container and then docker run it with proper arguments.

That would eliminate bash, templating, communicate, and code in a string literal...  Costs include another docker copy (not a big deal?) and the code being in two places (this really doesn't bother me).

Or, maybe we just replace the string/file with python code calling docker run for each command?  *shrug*  We could also walk away because it's working fine and we could always do this later.  :]